### PR TITLE
Cover Art Label & Cover Art Full Size composition with menu

### PR DESCRIPTION
**Pull request by fatihemreyildiz**
Hey everyone! 

I want to add cover art fetcher [[on this PR]](https://github.com/mixxxdj/mixxx/pull/4851), but for to do that I need to have another cover art label's without/different menu and DlgCoverArtFullSize. I have tried to make a base class without menu and inherit the existing label-full-size from that. But this caused as code duplication and new files which can make the code base crowded. Also, adding everything in a huge pr would make it difficult to review and also to merge. So, we wanted to move with small steps. Here is the first little PR of the cover art-fetcher.

To display cover arts without menu, I have tried composition approach, according to my tests, if there is no cover art menu passed to these two classes, it doesn't pop-up the menu on right click. In order for me to move forward, I need to find a best solution for to display cover art's with or without menu. 

Any feedback would be so good!

Thanks in advance.

### CODE REVIEW

File: src/library/dlgcoverartfullsize.cpp
Comment (daschuer) [inline src/library/dlgcoverartfullsize.cpp:16]: We use the stacked style of initialization list. Please revert to the style change (and help clang-format)

File: src/library/dlgtrackinfo.cpp
Comment (daschuer) [inline src/library/dlgtrackinfo.cpp:46]: This can also become make_parented<WCoverArtMenu>

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you.

File: src/library/dlgtrackinfo.cpp
Comment (fatihemreyildiz) [reply src/library/dlgtrackinfo.cpp:46]: Done! :+1:

File: src/widget/wcoverart.cpp
Comment (fatihemreyildiz) [inline src/widget/wcoverart.cpp:24]: I didn't do that intentionally this too, I think pre-commit made me do it in this way, is it correct or should it be like before?

File: src/library/dlgcoverartfullsize.cpp
Comment (daschuer) [inline src/library/dlgcoverartfullsize.cpp:34]: I think it is better to use here a common base class that has only the signals.

Comment (fatihemreyildiz) [reply src/library/dlgcoverartfullsize.cpp:34]: So Instead of QMenu should I use WCoverArtMenu in this case? For now I think only WCoverArtMenu can work so we can continue with the fetcher itself. A menu for fetcher can be implemented later on IMHO.

Wouldn't QObject casting work? Since it doesn't use RTTI, wouldn't it be faster and solve the issue regarding with the dynamic_casting? 

Or the problems with the casting it is not only the performance is it also technical stuff too?

Comment (daschuer) [reply src/library/dlgcoverartfullsize.cpp:34]: Yes technical, a QOject cast will also work and should indeed be preferred.

The main issue with upcasting, is that the expected child class is hidden in the implementation of the function body.

Comment (fatihemreyildiz) [reply src/library/dlgcoverartfullsize.cpp:34]: If i understand it correctly, we should go with the WCoverArtMenu instead of QMenu in the constructor right?

Comment (daschuer) [reply src/library/dlgcoverartfullsize.cpp:34]: Yes, that was my idea. We can alter the code later again if we introduce a second menu class.

Comment (fatihemreyildiz) [reply src/library/dlgcoverartfullsize.cpp:34]: Thank you for explaining, that's fair enough in the first place as you mentioned, with the second menu class it can be changed :+1:

Comment (fatihemreyildiz) [reply src/library/dlgcoverartfullsize.cpp]: Done :+1: I tested it with label without menu, it worked fine.

File: src/library/dlgtrackinfo.cpp
Comment (daschuer) [inline src/library/dlgtrackinfo.cpp:236]: here you can directly connect the menu and remove the connection and signals from the label.
If this is the only dependency the WCoverArtLabel can take the generic menu.

Comment (fatihemreyildiz) [reply src/library/dlgtrackinfo.cpp:236]: I actually tried it before, as far as I remember that was causing a bug. I click on the cover art layout, full size populated, after right click - clear cover was just clearing the cover art for the DlgTrackInfo, not in all of them which was not like in the original use-case.

Comment (daschuer) [reply src/library/dlgtrackinfo.cpp:236]: Can you try this again? I see no reason that it can cause a bug, because it is only a signal signal connection. without it we remove a redirection loop, that's all. (make sure to change the pointer AND the base class of the connected signal function pointer)

Comment (fatihemreyildiz) [reply src/library/dlgtrackinfo.cpp]: I deleted the signal signal connection, and it works just fine, but whenever I want to pass the QMenu, i can not because they have dependency.

WCoverArtLabel takes WCoverArtMenu and DlgCoverArtFullSize, DlgCoverArtFullSize also take WCoverArtMenu, so they have highly related dependency.

**Comment by daschuer** [comment]
Thank you. I think we can go out of draft state and ask for a second pair of unbiased eyes for review.

**Comment by fatihemreyildiz** [comment]
Done :+1: Thank you for the help and the review!

**Review by daschuer — APPROVED** [review]

**Comment by daschuer** [comment]
This LGTM.

@fatihemreyildiz can you rebase out the experimental commits?

**Comment by ronso0** [comment]
(this still needs a 2nd approval : )

I can take a look at the UI in the next days.

**Comment by fatihemreyildiz** [comment]
> This LGTM.
> 
> @fatihemreyildiz can you rebase out the experimental commits?

Thank you, Done, all clear :+1: . All experimental commits are out and just one left.

**Comment by ronso0** [comment]
> (this still needs a 2nd approval : )
> 
> I can take a look at the UI in the next days.

Oh, when I wrote that I must have confused this PR with the cover art fetcher.
This one doesn't change/add UI, "only" code changes, and for that my expertise is not sufficient :grimacing:
or did I miss something?
I guess someone else needs to take a look.

**Comment by daschuer** [comment]
I already did a detailed review. We just need a second pair of eyes that I have not missed something obvious, because I am probably biased. @ronso0 so if there is nothing wrong from you perspective I would be happy about a merge so that we can start to integrate it in the new MusicBrainz dialog. Thank you.

File: src/library/dlgcoverartfullsize.cpp
Comment (ronso0) [inline src/library/dlgcoverartfullsize.cpp:228]: This means if there is no menu instance the cover window can not be dragged anymore.
How about this:
```suggestion
    if (event->button() != Qt::LeftButton) {
        return;
    }
    if ((m_pCoverMenu != nullptr && !m_pCoverMenu->isVisible()) ||
            m_pCoverMenu == nullptr) {
```

File: src/widget/wcoverartlabel.h
Comment (ronso0) [inline src/widget/wcoverartlabel.h:4]: not required, please remove

**Review by ronso0 — CHANGES_REQUESTED** [review]

File: src/widget/wcoverartlabel.h
Comment (fatihemreyildiz) [reply src/widget/wcoverartlabel.h]: deleted.

File: src/library/dlgcoverartfullsize.cpp
Comment (fatihemreyildiz) [reply src/library/dlgcoverartfullsize.cpp]: I didn't notice that, I tested it before the requested changes, when there was no menu label, left click wasn't working probably and wasn't able to turn off the full size. With this change now it can. Thank you! :)

**Review by ronso0 — APPROVED** [review]

**Comment by ronso0** [comment]
Alright, LGTM*

**Comment by fatihemreyildiz** [comment]
Thank you for the feedback and reviews @daschuer and @ronso0 :octopus:

**Pull request merged**
